File: neet_code_all_cpp/sliding_window/contain_duplicate_2.cpp
#include <iostream>
#include <utility>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <climits>
#include <algorithm>

class Solution {
    public:
        // AC
        bool containsNearbyDuplicate(std::vector<int>& nums, int k) {
            std::unordered_map<int, std::vector<int>> map;

            // {value: index}なmapを作る
            for(int i = 0; i < (int)nums.size(); ++i) {
                if(map.find(nums[i]) != map.end()) {
                    map[nums[i]].push_back((int)i);
                } else {
                    map.insert({nums[i], std::vector{(int)i}});
                }
            }

            int min_diff = INT_MAX;
            for(auto& [v, idxs]: map) {
                if(idxs.size() > 1) {
                    std::sort(idxs.begin(), idxs.end());
                    for(int i = 0; i < (int)idxs.size() - 1; ++i) {
                        min_diff = std::min(min_diff, idxs[i + 1] - idxs[i]);
                    }
                }
            }

            return min_diff <= k;
        }
};

// 模範解答
class SolutionAns {
    public:
        bool containsNearbyDuplicate(std::vector<int>& nums, int k) {
            std::unordered_map<int, int> number_map;

            for(int i = 0; i < nums.size(); ++i) {
                auto num = nums[i];

                if(number_map.find(num) != number_map.end() && i - number_map[num] <= k) {
                    return true;
                } else {
                    number_map[num] = i;
                }
            }

            return false;
        }
};

// AC
// Rustの模範解答より
class SolutionAnsRust {
    public:
        bool containsNearbyDuplicate(std::vector<int>& nums, int k) {
            std::unordered_set<int> window;
            int l = 0;

            for(int r = 0; r < nums.size(); ++r) {
                if(r - l > k) {
                    window.erase(nums[l]);
                    ++l;
                }

                if(window.find(nums[r]) != window.end()) {
                    return true;
                }

                window.insert(nums[r]);
            }

            return false;
        }
};

int main(void) {
    using Case = std::pair<std::vector<int>, int>;

    Case case_1 = {{1,2,3,1}, 3};
    // => true
    Case case_2 = {{1,0,1,1}, 1};
    // => true
    Case case_3 = {{1,2,3,1,2,3}, 2};
    // => false
    Case case_4 = {{1} ,1};
    // => false

    Solution s_1;
    std::cout << s_1.containsNearbyDuplicate(case_1.first, case_1.second) << std::endl;
    std::cout << s_1.containsNearbyDuplicate(case_2.first, case_2.second) << std::endl;
    std::cout << s_1.containsNearbyDuplicate(case_3.first, case_3.second) << std::endl;
    std::cout << s_1.containsNearbyDuplicate(case_4.first, case_4.second) << std::endl;

    SolutionAns s_ans;
    std::cout << s_ans.containsNearbyDuplicate(case_1.first, case_1.second) << std::endl;
    std::cout << s_ans.containsNearbyDuplicate(case_2.first, case_2.second) << std::endl;
    std::cout << s_ans.containsNearbyDuplicate(case_3.first, case_3.second) << std::endl;
    std::cout << s_ans.containsNearbyDuplicate(case_4.first, case_4.second) << std::endl;

    SolutionAnsRust s_rs;
    std::cout << s_rs.containsNearbyDuplicate(case_1.first, case_1.second) << std::endl;
    std::cout << s_rs.containsNearbyDuplicate(case_2.first, case_2.second) << std::endl;
    std::cout << s_rs.containsNearbyDuplicate(case_3.first, case_3.second) << std::endl;
    std::cout << s_rs.containsNearbyDuplicate(case_4.first, case_4.second) << std::endl;
}

File: neet_code_all/src/sliding_window/contain_duplicate_2.rs
use std::collections::{HashMap, HashSet};

struct Solution {}
impl Solution {
    // AC
    pub fn contain_nearby_duplication(nums: Vec<i32>, k: i32) -> bool {
        let mut map: HashMap<i32, Vec<i32>> = HashMap::new();

        // {value: index} なmap
        for (idx, v) in nums.iter().enumerate() {
            if let Some(map_key) = map.get_mut(v) {
                map_key.push(idx as i32);
            } else {
                map.insert(*v, vec![idx as i32]);
            }
        }

        let mut min_diff = i32::MAX;
        for (_v, mut indxs) in map {
            //println!("idxs: {:?}", indxs);
            if indxs.len() > 1 {
                indxs.sort();
                for i in 0..indxs.len() - 1 {
                    min_diff = min_diff.min(indxs[i + 1] - indxs[i]);
                }
            }
        }
        //println!("min_diff: {}", min_diff);

        min_diff as i32 <= k
    }
}

// 模範解答
struct SolutionAns {}
impl SolutionAns {
    pub fn contain_nearby_duplication(nums: Vec<i32>, k: i32) -> bool {
        let mut window = HashSet::new();
        let mut l = 0;

        for (r, &num) in nums.iter().enumerate() {
            if r as i32 - l as i32 > k {
                window.remove(&nums[l]);
                l += 1;
            }

            if window.contains(&num) {
                return true;
            }

            window.insert(num);
        }

        false
    }
}

// C++の模範解答より
struct SolutionAnsCpp {}
impl SolutionAnsCpp {
    // AC
    pub fn contain_nearby_duplication(nums: Vec<i32>, k: i32) -> bool {
        let mut number_map: HashMap<i32, i32> = HashMap::new();

        for i in 0..nums.len() {
            let num = nums[i];

            if let Some(number_map_key) = number_map.get_mut(&num) {
                if i as i32 - *number_map_key <= k {
                    return true;
                } else {
                    *number_map_key = i as i32;
                }
            } else {
                number_map.insert(num, i as i32);
            }
        }

        false
    }
}

fn main() {
    let case_1 = (vec![1, 2, 3, 1], 3);
    // => true
    let case_2 = (vec![1, 0, 1, 1], 1);
    // => true
    let case_3 = (vec![1, 2, 3, 1, 2, 3], 2);
    // => false
    let case_4 = (vec![1], 1);
    // => false

    println!(
        "case_1: {}",
        Solution::contain_nearby_duplication(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {}",
        Solution::contain_nearby_duplication(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {}",
        Solution::contain_nearby_duplication(case_3.0.clone(), case_3.1)
    );
    println!(
        "case_4: {}",
        Solution::contain_nearby_duplication(case_4.0.clone(), case_4.1)
    );

    println!(
        "case_1: {}",
        SolutionAns::contain_nearby_duplication(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {}",
        SolutionAns::contain_nearby_duplication(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {}",
        SolutionAns::contain_nearby_duplication(case_3.0.clone(), case_3.1)
    );
    println!(
        "case_4: {}",
        SolutionAns::contain_nearby_duplication(case_4.0.clone(), case_4.1)
    );

    println!(
        "case_1: {}",
        SolutionAnsCpp::contain_nearby_duplication(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {}",
        SolutionAnsCpp::contain_nearby_duplication(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {}",
        SolutionAnsCpp::contain_nearby_duplication(case_3.0.clone(), case_3.1)
    );
    println!(
        "case_4: {}",
        SolutionAnsCpp::contain_nearby_duplication(case_4.0.clone(), case_4.1)
    );
}
